### DISCUSSION TITLE
Lesson 11 Mood NFT blank in Metamas

### DISCUSSION THREAD

**Discussion by PitrikYan** [Cyfrin/foundry-full-course-cu#1289] [Q&A]
When deployed contract on sepolia testnet, minted nft and tried to import it to Metamask account, it just shows up as a blank image. 
Everything done correctly, mood flipping working, token and image URIs are generated right, it can be displayed in browser. Just in Metamask it shows like on picture.
Please help :)

here is link to contract on etherscan:
https://sepolia.etherscan.io/address/0x124da0a6706de43e958cd8b9fab0d60a3215ce88

![metamask-blank-nft](https://github.com/Cyfrin/foundry-full-course-f23/assets/102188799/6070deff-b9b2-40ed-ac3c-26e770265311)

**Comment by azanux** [comment]
Hi @PitrikYan  it seems that your TokenURI is not well formatted , you can check your format here if you still have an error after this fix : https://jsonformatter.curiousconcept.com/#

your error is coming from your function tokenURI() #

Change the place of ":", you should have   abi.encodePacked(
                            `    '{"name":"',`
instead of 

abi.encodePacked(
                                `'{"name:" "',`

in your code : 
`return
            string(
                abi.encodePacked(
                    _baseURI(),
                    Base64.encode(
                        bytes(
                            abi.encodePacked(
                                '{"name:" "',
                                name(),
                                '", "description": "NFT which reflects owners mood", "image": "',
                                imageURI,
                                '", "attributes": [ { "trait_type": "mood", "value": 100 } ] }'
                            )
                        )
                    )
                )
            );
    }`

![Capture d’écran 2024-01-06 à 21 13 47](https://github.com/Cyfrin/foundry-full-course-f23/assets/7221893/b55ee9d3-a240-4c18-883a-b7a3392000d6)

**Reply by PitrikYan** [reply]
Thanks a lot for answer. Now I have JSON of tokenURI formatted well, but still has same issue with blank image in metamask, after I deployed new contract and minted new NFT, deleted the old one and imported new to Metamask.. So really dont know where the problem is now.

**Reply by azanux** [reply]
could you give me address of your new NFT contract ?

**Reply by danail-vasilev** [reply]
I have the same issue with metamask not showin the nft image. Tries some settings in metamask but to no avail

**Reply by azanux** [reply]
what's your nft address ?

**Comment by qjian33** [comment]
make sure json file is valid,and the format uri is like ipfs://QmdfhNKHBhQbfgh1W2jNZ3EESggALkSzGoKmzu4U8s6ZPL, could be the format uri problem

**Reply by qjian33** [reply]
and the image url is also that format

**Comment by EngrPips** [comment]
Interesting

**Comment by s3bc40** [comment]
So I finally managed to get the mood NFT showing up on Metamask ! It was long because I fought a little with : 

- JSON formating (comapring with Opensea format) 
- Metamask parameters

### For JSON
Comparing with NFT's on Opensea, I understood that I was missing some `"` and you also have to put it around the keys :
```json
{
  "name": "Mood NFT",
  "description": "An NFT that reflects your mood!",
  "attributes": [
    {
      "trait_type": "Mood",
      "value": 100
    }
  ],
  "image": "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAyNHB4IiBoZWlnaHQ9IjEwMjRweCIgdmlld0JveD0iMCAwIDEwMjQgMTAyNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KICA8cGF0aCBmaWxsPSIjMzMzIiBkPSJNNTEyIDY0QzI2NC42IDY0IDY0IDI2NC42IDY0IDUxMnMyMDAuNiA0NDggNDQ4IDQ0OCA0NDgtMjAwLjYgNDQ4LTQ0OFM3NTkuNCA2NCA1MTIgNjR6bTAgODIwYy0yMDUuNCAwLTM3Mi0xNjYuNi0zNzItMzcyczE2Ni42LTM3MiAzNzItMzcyIDM3MiAxNjYuNiAzNzIgMzcyLTE2Ni42IDM3Mi0zNzIgMzcyeiIvPgogIDxwYXRoIGZpbGw9IiNFNkU2RTYiIGQ9Ik01MTIgMTQwYy0yMDUuNCAwLTM3MiAxNjYuNi0zNzIgMzcyczE2Ni42IDM3MiAzNzIgMzcyIDM3Mi0xNjYuNiAzNzItMzcyLTE2Ni42LTM3Mi0zNzItMzcyek0yODggNDIxYTQ4LjAxIDQ4LjAxIDAgMCAxIDk2IDAgNDguMDEgNDguMDEgMCAwIDEtOTYgMHptMzc2IDI3MmgtNDguMWMtNC4yIDAtNy44LTMuMi04LjEtNy40QzYwNCA2MzYuMSA1NjIuNSA1OTcgNTEyIDU5N3MtOTIuMSAzOS4xLTk1LjggODguNmMtLjMgNC4yLTMuOSA3LjQtOC4xIDcuNEgzNjBhOCA4IDAgMCAxLTgtOC40YzQuNC04NC4zIDc0LjUtMTUxLjYgMTYwLTE1MS42czE1NS42IDY3LjMgMTYwIDE1MS42YTggOCAwIDAgMS04IDguNHptMjQtMjI0YTQ4LjAxIDQ4LjAxIDAgMCAxIDAtOTYgNDguMDEgNDguMDEgMCAwIDEgMCA5NnoiLz4KICA8cGF0aCBmaWxsPSIjMzMzIiBkPSJNMjg4IDQyMWE0OCA0OCAwIDEgMCA5NiAwIDQ4IDQ4IDAgMSAwLTk2IDB6bTIyNCAxMTJjLTg1LjUgMC0xNTUuNiA2Ny4zLTE2MCAxNTEuNmE4IDggMCAwIDAgOCA4LjRoNDguMWM0LjIgMCA3LjgtMy4yIDguMS03LjQgMy43LTQ5LjUgNDUuMy04OC42IDk1LjgtODguNnM5MiAzOS4xIDk1LjggODguNmMuMyA0LjIgMy45IDcuNCA4LjEgNy40SDY2NGE4IDggMCAwIDAgOC04LjRDNjY3LjYgNjAwLjMgNTk3LjUgNTMzIDUxMiA1MzN6bTEyOC0xMTJhNDggNDggMCAxIDAgOTYgMCA0OCA0OCAwIDEgMC05NiAweiIvPgo8L3N2Zz4="
}
``` 

You can use https://jsonformatter.org/ to check if there is anything missing

---

### For Metamask

Even though it was fixed with the JSON format, whenever I wanted to deploy the contract and mint : 
- or I was having a "Not permited" error when I was the owner of the NFT
- or the image wouldn't show up 

Strange since everytime, I was deploying over and over again anvil with a fresh contract. And checking Metamask, I saw that I was not having any loader even if I was closing Anvil. The only way to have this was to : 
1. close anvil
2. remove the local network from metamask
3. add it again and switch to it (with anvil running)

I also cliked on `Setting/Advanced/Clear activity tab data`, and enabled my local network to `Settings/Security & privacy/Show incoming transactions`.

After all of this I got my NFT, even after a flip mood : 
![image](https://github.com/user-attachments/assets/acaf846f-5df2-4479-8aa0-58c86509adfe)

I have a strange feeling of cache data or something like this. I do not have the right origin of this, maybe on my side (the browser ? human error ?), but if it can help :+1:

**Reply by EngrPips** [reply]
Interesting, That's a lot of work. well done.

**Reply by Marvelousy20** [reply]
How can you possibly switch to it without having anvil running locally?

**Reply by EngrPips** [reply]
hmmm 🤔

**Reply by s3bc40** [reply]
> How can you possibly switch to it without having anvil running locally?

Of course Anvil is running when I make the transactions, my bad in the explanation. I have just seen that in the parenthesis...

To switch you have to do the same as Patrick : delete it in metamask -> do the flip -> import it again to see the change.

The explanation before is just testing why I was not having the image in the first place. So I have restarted anvil a lot of time, and inspected the metamask params 😌

**Reply by s3bc40** [reply]
Fixed in the main post thanks : anvil should run 👍

**Reply by Marvelousy20** [reply]
> > How can you possibly switch to it without having anvil running locally?
> 
> Of course Anvil is running when I make the transactions, my bad in the explanation. I have just seen that in the parenthesis...
> 
> To switch you have to do the same as Patrick : delete it in metamask -> do the flip -> import it again to see the change.
> 
> The explanation before is just testing why I was not having the image in the first place. So I have restarted anvil a lot of time, and inspected the metamask params 😌

Thanks. I am still having the same issue tho. I have removed and re added anvil but to no avail. The NFT is still blank.

**Reply by EngrPips** [reply]
Have you also enabled all `nft display` settings in metamask settings?

**Reply by Marvelousy20** [reply]
> Have you also enabled all `nft display` settings in metamask settings?

This fixed it. Thank you so much. Appreciate it.

**Reply by s3bc40** [reply]
> Have you also enabled all `nft display` settings in metamask settings?

Forgot about this param I also enabled it while fixing. Great 🎉

**Reply by EngrPips** [reply]
> > Have you also enabled all `nft display` settings in metamask settings?
> 
> This fixed it. Thank you so much. Appreciate it.

Awesome, Keep crushing it.
